runtime/tests/jvmtitests/src/com/ibm/jvmti/tests/fieldwatch/fw001.c

```diff
@@ -0,0 +1,227 @@
+
+JNIEXPORT jint JNICALL
+Java_com_ibm_jvmti_tests_fieldwatch_fw001_endTest(JNIEnv *jni_env, jobject cls)
+{
+	(*jni_env)->ReleaseStringUTFChars(jni_env, testClassString, testClassName);
+	return JNI_OK;
+}
+
+/* Add or remove field watches. */
+JNIEXPORT jint JNICALL
+Java_com_ibm_jvmti_tests_fieldwatch_fw001_modifyWatches(JNIEnv *env, jclass rcv, jstring className, jstring fieldName, jstring fieldSignature, jboolean isStatic, jboolean isRead, jboolean isWrite, jboolean isAdd)
+{
+	const char * _fieldName = (*env)->GetStringUTFChars(env, fieldName, NULL);
+	const char * _fieldSignature = (*env)->GetStringUTFChars(env, fieldSignature, NULL);
+	const char * _className = (*env)->GetStringUTFChars(env, className, NULL);
+	jfieldID fid;
+
+	if (NULL == _className) {
+		printf("_className string is NULL\n");
+		return JNI_ERR;
+	}
+
+	if (NULL == _fieldName) {
+		printf("_fieldName string is NULL\n");
+		return JNI_ERR;
+	}
+
+	if (NULL == _fieldSignature) {
+		printf("_fieldSignature string is NULL\n");
+		return JNI_ERR;
+	}
+
+	rcv = (*env)->FindClass(env, _className);
+	if (NULL == rcv) {
+		printf("Unable to find class with name: %s\n", _className);
+		return JNI_ERR;
+	}
+
+	if (isStatic) {
+		fid = (*env)->GetStaticFieldID(env, rcv, _fieldName, _fieldSignature);
+	} else {
+		fid = (*env)->GetFieldID(env, rcv, _fieldName, _fieldSignature);
+	}
+
+	if (NULL != fid) {
+		if (isRead) {
+			jvmtiError err = JVMTI_ERROR_NONE;
+			if (isAdd) {
+				err = (*globalEnv)->SetFieldAccessWatch(globalEnv, rcv, fid);
+			} else {
+				err = (*globalEnv)->ClearFieldAccessWatch(globalEnv, rcv, fid);;
+			}
+
+			if (JVMTI_ERROR_NONE != err) {
+				printf( "**Error**: Failed to %s field access watch with JVMTI Error: %d\n", isAdd ? "add" : "remove",  err);
+				return JNI_ERR;
+			}
+		}
+
+		if (isWrite) {
+			jvmtiError err = JVMTI_ERROR_NONE;
+			if (isAdd) {
+				err = (*globalEnv)->SetFieldModificationWatch(globalEnv, rcv, fid);
+			} else {
+				err = (*globalEnv)->ClearFieldModificationWatch(globalEnv, rcv, fid);
+			}
+
+			if (JVMTI_ERROR_NONE != err) {
+				printf( "**Error**: Failed to %s field modification watch with JVMTI Error Code: %d\n", isAdd ? "add" : "remove",  err);
+				return JNI_ERR;
+			}
+		}
+	} else {
+		printf("**ERROR**: Unable to modify fieldwatch(%s%s) for %s. Couldn't find a valid fieldID associated with class: %s, field: %s, signature: %s, isStatic: %s, modificationType: %s\n",
+													isRead ? "access" : "", isWrite ? ", modification" : "", _fieldName, _className, _fieldName, _fieldSignature, isStatic ? "Yes" : "No", isAdd ? "add" : "remove");
+		return JNI_ERR;
+	}
+
+	printf("**NOTE**: Modified fieldwatch(%s%s) for %s. The fieldID is: %p and it is associated with class: %s, field: %s, signature: %s, isStatic: %s, modificationType: %s\n\n",
+													isRead ? "access" : "", isWrite ? ", modification" : "", _fieldName, fid, _className, _fieldName, _fieldSignature, isStatic ? "Yes" : "No", isAdd ? "add" : "remove");
```

Comment:
The formatting here is a bit odd - for really long things like this, I tend to do something like:
```
printf("string",
   parm1,
   parm2)
```

or

```
printf("string"
   , parm1
   , parm2);
```

or some variant like that.

Reply:
Is this 624a4e94e4 better?

Reply:
Sorry, I just don't know how to properly embed a tab in the git text. Yes, it's much easier to read now, thanks.